### PR TITLE
Support Iterator for all Listings

### PR DESCRIPTION
**Pull request by BlackbitNeueMedien**
This PR harmonizes all descendants of `Pimcore\Model\Listing\AbstractListing`, so all implement the `\Iterator` interface.
Originally I that that was meant with "For all Listing classes it's not necessary anymore to call the load() method before calling the actual getter for the items" in https://pimcore.com/docs/6.x/Development_Documentation/Installation_and_Upgrade/Updating_Pimcore/V5_to_V6.html#page_General-Changes but in the code a lot of Listing classes did not implement `\Iterator` yet.

This PR actually changes some setter return values. If you do not want to harmonize those for backwards compatibility reasons, I will remove the `return`s there to have the `void` return type they had before.

### CODE REVIEW

File: bundles/EcommerceFrameworkBundle/CartManager/Cart/Listing.php [hunk 1]
```diff
@@ -23,10 +23,6 @@
  */
 class Listing extends \Pimcore\Model\Listing\AbstractListing
 {
-    /**
-     * @var array
-     */
-    public $carts;
```
Comment (dpfaffenbauer) [inline bundles/EcommerceFrameworkBundle/CartManager/Cart/Listing.php]: BC break

File: bundles/EcommerceFrameworkBundle/CartManager/CartCheckoutData/Listing.php [hunk 2]
```diff
@@ -23,10 +23,6 @@
  */
 class Listing extends \Pimcore\Model\Listing\AbstractListing
 {
-    /**
-     * @var array
-     */
-    public $cartCheckoutDataItems;
```
Comment (dpfaffenbauer) [inline bundles/EcommerceFrameworkBundle/CartManager/CartCheckoutData/Listing.php]: BC break

File: bundles/EcommerceFrameworkBundle/CartManager/CartItem/Listing.php [hunk 3]
```diff
@@ -25,10 +25,6 @@
  */
 class Listing extends \Pimcore\Model\Listing\AbstractListing
 {
-    /**
-     * @var array
-     */
-    public $cartItems;
```
Comment (dpfaffenbauer) [inline bundles/EcommerceFrameworkBundle/CartManager/CartItem/Listing.php]: BC break

File: lib/Model/Listing/AbstractListing.php [hunk 4]
```diff
@@ -445,4 +447,65 @@ public function getConditionVariablesFromSetCondition()
     {
         return $this->conditionVariablesFromSetCondition;
     }
+
+    public function getData(): ?array
```
Comment (dpfaffenbauer) [inline lib/Model/Listing/AbstractListing.php]: BC break, don't set the return value, what if someone else extends from AbstractListing?

File: models/DataObject/Listing/Concrete.php [hunk 5]
```diff
@@ -54,7 +54,7 @@ abstract class Concrete extends Model\DataObject\Listing
      */
     public function __construct()
     {
-        $this->initDao('\\Pimcore\\Model\\DataObject\\Listing\\Concrete');
+        $this->initDao(__CLASS__);
```
Comment (dpfaffenbauer) [inline models/DataObject/Listing/Concrete.php]: not sure if right, but doesn't ```__CLASS__``` return the actual class also on classes that extend Listing\Concrete?

File: lib/Model/Listing/AbstractListing.php (hunk 4 above)
Comment (BlackbitNeueMedien) [reply lib/Model/Listing/AbstractListing.php]: Ok, changed this as it is not really necessary.

File: bundles/EcommerceFrameworkBundle/CartManager/Cart/Listing.php (hunk 1 above)
Comment (BlackbitNeueMedien) [reply bundles/EcommerceFrameworkBundle/CartManager/Cart/Listing.php]: Yes it is. Perhaps we should leave classes with public item properties as they are. But as most other class variables are protected meanwhile we have to tackle this any time...

File: models/DataObject/Listing/Concrete.php (hunk 5 above)
Comment (BlackbitNeueMedien) [reply models/DataObject/Listing/Concrete.php]: No, __CLASS__ always points to the class where this statement is in.

File: lib/Model/Listing/AbstractListing.php (hunk 4 above)
Comment (BlackbitNeueMedien) [reply lib/Model/Listing/AbstractListing.php]: But as this function did not exist before this PR I actually do not see a BC break.

Comment (dpfaffenbauer) [reply lib/Model/Listing/AbstractListing.php]: but in many other Listings it did

Comment (BlackbitNeueMedien) [reply lib/Model/Listing/AbstractListing.php]: I only found it in https://github.com/pimcore/pimcore/blob/bb4fd386fd09d3e5f4e26041f7aef2eccae0aa7e/models/Notification/Listing.php#L52

But anyway: The return type hint does not add any value imho, so I am fine as it is now (without return type hint)

File: bundles/EcommerceFrameworkBundle/CartManager/Cart/Listing.php (hunk 1 above)
Comment (BlackbitNeueMedien) [reply bundles/EcommerceFrameworkBundle/CartManager/Cart/Listing.php]: Resolved

File: bundles/EcommerceFrameworkBundle/CartManager/CartCheckoutData/Listing.php (hunk 2 above)
Comment (BlackbitNeueMedien) [reply bundles/EcommerceFrameworkBundle/CartManager/CartCheckoutData/Listing.php]: Resolved

File: bundles/EcommerceFrameworkBundle/CartManager/CartItem/Listing.php (hunk 3 above)
Comment (BlackbitNeueMedien) [reply bundles/EcommerceFrameworkBundle/CartManager/CartItem/Listing.php]: Resolved

**Comment by BlackbitNeueMedien** [comment]
Now it works without BC breaks.
And also IDEs recognize the return type automatically via `@method >ModelClass> current()` in all Listing classes.

File: bundles/EcommerceFrameworkBundle/CartManager/Cart/Listing.php [hunk 6]
```diff
@@ -62,6 +61,6 @@ public function getCarts()
      */
     public function setCarts($carts)
     {
-        $this->carts = $carts;
+        return $this->setData($carts);
```
Comment (NiklasBr) [inline bundles/EcommerceFrameworkBundle/CartManager/Cart/Listing.php]: No longer `@return void`

File: models/DataObject/QuantityValue/Unit/Listing.php [hunk 7]
```diff
@@ -49,18 +55,14 @@ public function isValidOrderKey($key)
...
+        return $this->getData();
     }
 
     /**
      * @param array $units
      */
     public function setUnits($units)
     {
-        $this->units = $units;
+        return $this->setData($units);
```
Comment (NiklasBr) [inline models/DataObject/QuantityValue/Unit/Listing.php]: Add `@return` in Doc?

File: models/ImportConfig/Listing.php [hunk 8]
```diff
@@ -30,23 +31,24 @@ class Listing extends Model\Listing\AbstractListing
...
+        return $this->getData();
     }
 
     /**
      * @param array $importconfigs
      */
     public function setImportconfigs(array $importconfigs)
     {
-        $this->importconfigs = $importconfigs;
+        return $this->setData($importconfigs);
```
Comment (NiklasBr) [inline models/ImportConfig/Listing.php]: Add `@return ...` to Doc

**Comment by BlackbitNeueMedien** [comment]
@NiklasBr That i what I meant with "This PR actually changes some setter return values. If you do not want to harmonize those for backwards compatibility reasons, I will remove the returns there to have the void return type they had before."
When the Pimcore maintainers want to syncronize the Listings, I will update the return doctypes. Otherwise I will remove the `return` in the setters which had `void` return type till now. The latter would provide full BC compatibility.

**Comment by NiklasBr** [comment]
@BlackbitNeueMedien What I requested was not either that or the other. I wanted to `@return` declaration *in* the Doc to actually reflect the reality. Whichever you choose to make the reality.

This is not correct:
```php
/**
 * @return void
 */
function exampleFunction() {
      return [];
}
```

This is:
```php
/**
 * @return array
 */
function exampleFunction() {
      return [];
}
```

And it has nothing to do with BC compatibility.

File: bundles/EcommerceFrameworkBundle/CartManager/Cart/Listing.php (hunk 6 above)
Comment (BlackbitNeueMedien) [reply bundles/EcommerceFrameworkBundle/CartManager/Cart/Listing.php]: Fixed

File: models/ImportConfig/Listing.php (hunk 8 above)
Comment (BlackbitNeueMedien) [reply models/ImportConfig/Listing.php]: Fixed

**Comment by BlackbitNeueMedien** [comment]
Sorry @NiklasBr, you are right. Now return doctypes are corrected. 
But my original statement concerning the decision whether to harmonize all Listing setter methods with returning the Listing object at the cost of a slight BC creak remains. Waiting for a reply of the maintainers. If we want to keep BC, I will remove the `return` from those setter methods which had `@return void` before.

File: bundles/EcommerceFrameworkBundle/VoucherService/Token/Listing.php [hunk 9]
```diff
@@ -20,13 +20,19 @@
...
 /**
  * @method Token[] load()
+ * @method Token current()
  * @method int getTotalCount()
  * @method \Pimcore\Bundle\EcommerceFrameworkBundle\VoucherService\Token\Listing\Dao getDao()
  */
-class Listing extends \Pimcore\Model\Listing\AbstractListing implements \Iterator, AdapterInterface, AdapterAggregateInterface
+class Listing extends \Pimcore\Model\Listing\AbstractListing implements AdapterInterface, AdapterAggregateInterface
 {
     public $tokens;
```
Comment (brusch) [inline bundles/EcommerceFrameworkBundle/VoucherService/Token/Listing.php]: I guess we should mark those "replaced" properties in the listing classes as `@deprecated` to that we can remove them in v7.

**Comment by brusch** [comment]
@BlackbitNeueMedien All the rest looks good - great work! Just waiting for the feedback to my previous comment - then it's ready to be merged 🤓
@NiklasBr @dpfaffenbauer Thanks for your valuable feedback/review!

File: bundles/EcommerceFrameworkBundle/VoucherService/Token/Listing.php (hunk 9 above)
Comment (BlackbitNeueMedien) [reply bundles/EcommerceFrameworkBundle/VoucherService/Token/Listing.php]: Done

**Comment by BlackbitNeueMedien** [comment]
@brusch One last thing to mention: `\Pimcore\Model\Dao\AbstractDao` actually does not define a `load()` method. We also cannot add it because this Dao does also get used for other things as Listings. So there is a slight risk of a BC break if someone extended `AbstractListing` and its Dao class does not have a `load()` method.
The only way to clean this up would be to introduce a new `ListingDao` class which extends `\Pimcore\Model\Dao\AbstractDao` and contains the `load()` method. But this would of course be also a BC break as `AbstractListing::getDao()` then would need to return a `ListingDao` object.

**Comment by brusch** [comment]
@BlackbitNeueMedien I'd value the risk as very low, however we can think about adding a `method_exists()` check before calling `$this->getDao()->load();` - the question is ... is it worth it? 
What do you think?

**Comment by dpfaffenbauer** [comment]
I am pretty sure people extend the `AbstractDao` and `AbstractListing` for custom Pimcore DAO Models. So, I'd go for a `method_exist` in Pimcore 6 and change that to an actual function in Pimcore 7.

**Comment by BlackbitNeueMedien** [comment]
I added the method_exists() call.
But nevertheless I would like to have a structure for new Listing classes. When somebody writes a new Listing class this dependency for a `load()` method of the Dao is nowhere explicitly defined. The mentioned `ListingDao` could define an abstract load method.
But perhaps this is something we should keep in mind for version 7...

**Comment by dpfaffenbauer** [comment]
definitely something for Pimcore 7, everything else would be a BC break.

**Comment by dpfaffenbauer** [comment]
you could trigger a notice when the load function doesn't exist for now, cause it is expected. Something like:

```
@trigger_error(
            'The DAO class expects a load method that currently wasn't declared in the AbstractClass. The method will be added in Pimcore 7,
            E_USER_DEPRECATED
        );
```

**Comment by BlackbitNeueMedien** [comment]
Done. With this deprecation error we will easily find this place when V7 gets prepared ;-)
Now I am fine with this PR.

**Comment by brusch** [comment]
@BlackbitNeueMedien just one more thing, can you please add the following code to `\Pimcore\Model\Listing\Dao\AbstractDao` : 

```php
    /**
     * @TODO abstract method to be activated in Pimcore 7
     * @return array
     */
    /*abstract public function load();*/
```

Then I'm also fine 😉
Thanks again!

**Pull request merged**